packages/block-library/src/heading/edit.js

```diff
@@ -28,14 +30,37 @@ function HeadingEdit( {
 	mergedStyle,
 	clientId,
 } ) {
-	const { textAlign, content, level, placeholder } = attributes;
+	const { textAlign, content, level, placeholder, anchor } = attributes;
 	const tagName = 'h' + level;
 	const blockProps = useBlockProps( {
 		className: classnames( {
 			[ `has-text-align-${ textAlign }` ]: textAlign,
 		} ),
 		style: mergedStyle,
 	} );
+	const allHeadingAnchors = useAllHeadingAnchors( clientId );
+
+	// Initially set anchor for headings that have content but no anchor set.
+	// This is used when transforming a block to heading, or for legacy anchors.
+	useEffect( () => {
+		if ( ! anchor && content ) {
+			setAttributes( {
+				anchor: generateAnchor( content, allHeadingAnchors ),
+			} );
+		}
+	}, [ allHeadingAnchors ] );
```

Comment:
We can't have `setAttributes` in the component as it produces side effects and that's why I used `useEffect`. We need the `allHeadingAnchors` dependency because the `setAttributes` is run async and `allHeadingAnchors` initially return nothing as there was no time to update the `anchors`.